File: algorithms/2_medium/M/00945_minimum_increment_to_make_array_unique/table.c
int minIncrementForUnique(const int *restrict nums, const int len) {
    if (nums == NULL || len <= 1) { return 0; }
    if (len > 100000) { return -1; }

    int max = INT_MIN;
    int min = INT_MAX;

    for (int i = 0; i < len; i += 1) {
        max = MAX(max, nums[i]);
        min = MIN(min, nums[i]);
    }

    const size_t width = (size_t)((int64_t)max - min + 1);
    uint32_t *restrict freqs = calloc(width, sizeof(*freqs));
    if (freqs == NULL) { return -1; }

    for (int i = 0; i < len; i += 1) {
        freqs[nums[i] - min] += 1;
    }

    uint32_t res = 0;
    for (size_t i = 0; i < width - 1; i += 1) {
        const uint32_t freq = freqs[i];
        if (freq > 1) {
            res += freq - 1;
            freqs[i] -= freq - 1;
            freqs[i + 1] += freq - 1;
        }
    }

    const uint64_t last = (uint64_t)(freqs[width - 1] - 1);
    res += (uint32_t)(last * (last + 1) / 2);

    free(freqs);
    return (int)res;
}

File: algorithms/2_medium/M/00945_minimum_increment_to_make_array_unique/table.rs
impl Solution {
    pub fn min_increment_for_unique(nums: Vec<i32>) -> i32 {
        if nums.len() <= 1 { return 0; }

        assert!(nums.len() <= 100_000);
        let mut max = i32::MIN;
        let mut min = i32::MAX;

        for &num in &nums {
            max = max.max(num);
            min = min.min(num);
        }

        let width = ((max as i64) - (min as i64) + 1) as usize;
        let mut freqs = vec![0_u32; width];
        for num in nums {
            freqs[(num - min) as usize] += 1;
        }

        let mut res = 0;
        for i in 0..(width - 1) {
            let freq = freqs[i];
            if freq > 1 {
                res += freq - 1;
                freqs[i] -= freq - 1;
                freqs[i + 1] += freq - 1;
            }
        }

        let last = (freqs[width - 1] - 1) as u64;
        res += (last * (last + 1) / 2) as u32;

        res.try_into().unwrap()
    }
}
